perf/pkeyread.c

```diff
@@ -0,0 +1,174 @@
+    0x04, 0x20           /*     OCTET STRING(32 bytes) */
+    /* key follows here */
+};
+
+#define X25519_PRIVATE_KEY_SZ	32	/* 32 bytes 256 bits */
+#define X25519_PUBLIC_KEY_SZ	32	/* 32 bytes 256 bits */
+#define PKEY_INFO_HDR_SZ	sizeof(PKEY_INFO_HDR)
+#define	PKEY_DER_SZ		(PKEY_INFO_HDR_SZ + X25519_PRIVATE_KEY_SZ)
+
+static int
+pkcs8_decode(const unsigned char *pdata)
+{
+    unsigned char pkey_der[PKEY_DER_SZ];
+    const unsigned char *pder = (const unsigned char *)pkey_der;
+    unsigned char *priv_data = pkey_der + PKEY_INFO_HDR_SZ;
+    const unsigned char *pk_buf = NULL;
+    int pk_buf_len;
+    PKCS8_PRIV_KEY_INFO *pkey_info8 = NULL;
+    EVP_PKEY *pkey = NULL;
+    X509_PUBKEY *key = NULL;
+    int err = 1;
+
+
+    memcpy(pkey_der, PKEY_INFO_HDR, PKEY_INFO_HDR_SZ);
+    memcpy(priv_data, pdata, X25519_PRIVATE_KEY_SZ);
+
+    pkey_info8 = d2i_PKCS8_PRIV_KEY_INFO(NULL, &pder, PKEY_DER_SZ);
+    if (pkey_info8 == 0)
+	goto error;
+
+    pkey = EVP_PKCS82PKEY(pkey_info8);
+    if (key == NULL)
+	goto error;
+
+    if (X509_PUBKEY_set(&key, pkey) == 0)
+	goto error;
+
+    if (X509_PUBKEY_get0_param(NULL, &pk_buf, &pk_buf_len, NULL, key) == 0)
+	goto error;
+
+    err = (pk_buf == NULL) || (pk_buf_len != X25519_PUBLIC_KEY_SZ);
+
+error:
+    X509_PUBKEY_free(key);
+    EVP_PKEY_free(pkey);
+    PKCS8_PRIV_KEY_INFO_free(pkey_info8);
+
+    return err;
+}
+
+static int 
+pkcs8_decode_batch(void)
+{
+    unsigned int i, j;
+    unsigned char pkey[X25519_PRIVATE_KEY_SZ];
+
+    memset(pkey, 0, X25519_PRIVATE_KEY_SZ);
+    for (i = 0; i < NUM_LOOPS; i++) {
+	pkey[0] = i;
+	for(j = 0 ; j < NUM_LOOPS; j++) {
+	    pkey[1] = j;
+	    if (pkcs8_decode(pkey) == 1) {
+		return 0;
+	    }
+	}
+    }
+
+    return 1;
+}
+
+static void
+do_pkcs8_decode(size_t unused)
+{
+    int i;
+    int err = 0;
+
+    for (i = 0; err == 0 && i < NUM_CALLS_PER_RUN / threadcount; i++)
+        err = pkcs8_decode_batch();
+
+    if (err)
+	fail = 1;
+}
+
+int
+main(int argc, const char * const argv[])
+{
+    OSSL_TIME duration;
+    uint64_t us;
+    double avcalltime;
+    int terse = 0;
+    int argnext;
+
+    if ((argc != 2 && argc != 3)
+                || (argc == 3 && strcmp("--terse", argv[1]) != 0)) {
+        printf("Usage: pemread [--terse] threadcount\n");
+        return EXIT_FAILURE;
+    }
+
+    if (argc == 3) {
+        terse = 1;
+        argnext = 2;
+    } else {
+        argnext = 1;
+    }
+
+    threadcount = atoi(argv[argnext]);
+    if (threadcount < 1) {
+        printf("threadcount must be > 0\n");
+        return EXIT_FAILURE;
+    }
+
+    if (!perflib_run_multi_thread_test(do_pkcs8_decode, threadcount, &duration)) {
+        printf("Failed to run the test\n");
+        return EXIT_FAILURE;
+    }
+
+    if (fail) {
+        printf("Error during test\n");
+        return EXIT_FAILURE;
+    }
+
+    us = ossl_time2us(duration);
+
+    avcalltime = (double)us / NUM_CALL_BLOCKS_PER_RUN;
```

Comment:
@t8m  and I debated this a while back.  While there is nothing specifically incorrect about doing the measurement this way, it leads to a behavior in which upping the thread count leads to decreasing per read times that are not particularly realistic (i.e. 100000 iterations performed by 4 threads completes in half the time that doing so with 2 threads does).  While accurate in total time, it doesn't really represent the average time per call to pkcs8_decode_batch.

Instead, the thought was to measure each threads execution time and average those numbers, see:
https://github.com/openssl/tools/pull/184

Reply:
Thanks Neil I took your changes to `pkeyread.c` The output with your changes in looks as follows:
```
lifty$ ./pkeyread -k all -f all -t 4 -v 
Average time per PEM_read_bio_PrivateKey(dh) calls: 1021.203000us
Average time per X509_PUBKEY_get0_param(dh) calls: 1057.180000us
Average time per PEM_read_bio_PrivateKey(dhx) calls: 247.702000us
Average time per X509_PUBKEY_get0_param(dhx) calls: 265.998000us
Average time per PEM_read_bio_PrivateKey(dsa) calls: 270.647000us
Average time per X509_PUBKEY_get0_param(dsa) calls: 313.164000us
Average time per PEM_read_bio_PrivateKey(ec) calls: 136.857000us
Average time per X509_PUBKEY_get0_param(ec) calls: 308.600000us
Average time per PEM_read_bio_PrivateKey(rsa) calls: 88.839000us
Average time per X509_PUBKEY_get0_param(rsa) calls: 159.471000us
Average time per PEM_read_bio_PrivateKey(xkey) calls: 98.997000us
Average time per X509_PUBKEY_get0_param(xkey) calls: 70.429000us
```

The pull request will be updated shortly.